README.md
Clone this repository.

```
git clone git@github.com:holatuwol/liferay-faster-deploy.git
```

# Use for Local Development

Add a Bash function to `.bash_aliases` (or the equivalent on whichever shell you're using) which calls the script while also specifying where you want the builds to go by setting the `BUILD_FOLDER_PREFIX` variable. Use whatever shorthand you think makes sense, with the example here being `rd`.

```
rd() {
	BUILD_FOLDER_PREFIX=/opt/liferay \
		BUILD_FOLDER_SUFFIX=bundles \
		TAG_ARCHIVE_MIRROR='http://cloud-10-50-0-165/builds/fixpacks' \
		BRANCH_ARCHIVE_MIRROR='http://cloud-10-50-0-165/builds/branches' \
			$HOME/Work/tooling/redeploy/redeploy
}
```

Navigate to the root of the portal repository and invoke the function you created.

```
cd /path/to/portal/source

rd
```

# Use for a Build Server

TODO

crontab/changelog
#!/bin/bash

getchangelog() {
	echo '    "'${1}'": ['

	git ls-tree -r --name-only upstream/${1} -- modules/ > .modules

	(
		for file in $(grep -Fh app.properties .modules); do
			getfixes $(dirname ${file}) ${1}
		done
	) | sed '$s/,$//g'

	rm -f .modules

	echo -n '    ]'
}

getcontenttargeting() {
	if [ ! -d ${HOME}/liferay-apps-content-targeting-ee ]; then
		return 0
	fi

	local LATEST_TAG=
	local LATEST_HASH=
	local LATEST_BRANCH=

	pushd ${HOME}/liferay-apps-content-targeting-ee > /dev/null
	git fetch upstream --no-tags

	if [ "7.0.x" == "${1}" ]; then
		LATEST_TAG=$(git tag | grep '^v2\.[0-9]*\.[0-9]*$' | sed 's/^v//g' | sort -n -t '.' -k1 | tail -1)
		LATEST_HASH=v${LATEST_TAG}
		LATEST_BRANCH='7.0.x/2.x'

		if [ "2.1.2" == "${LATEST_TAG}" ]; then
			LATEST_TAG=2.2.0
			LATEST_HASH=v2.2.0-rc2
		fi
	elif [ "7.1.x" == "${1}" ]; then
		LATEST_TAG=$(git tag | grep '^v3\.[0-9]*\.[0-9]*$' | sed 's/^v//g' | sort -n -t '.' -k1 | tail -1)
		LATEST_HASH=v${LATEST_TAG}
		LATEST_BRANCH='develop'
	else
		popd > /dev/null
		return 0
	fi

	echo -n '      {'
	echo -n '"marketplace": "43707691",'
	echo -n '"name": "Liferay Audience Targeting", '
	echo -n '"version": "'${LATEST_TAG}'", '
	echo -n '"path": "liferay/liferay-apps-content-targeting-ee", '
	echo -n '"branch": "'${LATEST_BRANCH}'", '

	echo -n '"releaseDate": "'$(git log -1 --pretty='%ci' ${LATEST_HASH} | awk '{ print $1 }')'", '

	echo -n '"fixes": ['

	local CHANGELOG=$(git log --pretty='%s' ${LATEST_HASH}..upstream/${LATEST_BRANCH} | grep -Eo '[A-Z]+-[0-9]+' | grep -vF 'LRDOCS' | grep -vF LRQA | grep -vF -- '-0' | sort -u)

	if [ "" != "${CHANGELOG}" ]; then
		echo -n "${CHANGELOG}" | awk '{ print "\"" $1 "\"" }' | tr '\n' ',' | sed 's/,$//g' | sed 's/,/, /g'
	fi

	echo -n ']'
	echo '},'

	popd > /dev/null
}

getfixes() {
	local APPBND_FOLDER=$(echo ${1} | sed 's@\.releng/@@g')
	local MODULE_BRANCH=${2}
	local LIFERAY_VERSION_NUMBER="$(echo ${MODULE_BRANCH} | grep -o '[0-9]' | perl -ne 'chomp and print')10"
	local APPBND_BRANCH=${MODULE_BRANCH}
	local MODULE_FOLDER=${APPBND_FOLDER}

	if [ "" == "$(grep -F ${APPBND_FOLDER}/app.bnd .modules)" ]; then
		return 0
	fi

	if [ "${APPBND_FOLDER}" != "${MODULE_FOLDER}" ]; then
		return 0
	fi

	if [ "" != "$(git show upstream/${APPBND_BRANCH}:${APPBND_FOLDER}/app.bnd | grep -F 'Liferay-Releng-Deprecated: true')" ]; then
		return 0
	fi

	if [ "" != "$(git show upstream/${APPBND_BRANCH}:${APPBND_FOLDER}/app.bnd | grep -F 'Liferay-Releng-Marketplace: false')" ]; then
		return 0
	fi

	if [ "" != "$(git show upstream/${APPBND_BRANCH}:${APPBND_FOLDER}/app.bnd | grep -F 'Liferay-Releng-Portal-Required: true')" ]; then
		return 0
	fi

	local NAME=$(git show upstream/${APPBND_BRANCH}:$1/app.properties | grep -F app.marketplace.title= | cut -d'=' -f 2)

	if [ "" == "${NAME}" ]; then
		return 0
	fi

	local ID=$(git show upstream/${APPBND_BRANCH}:$1/app.properties | grep -F app.marketplace.id= | cut -d'=' -f 2)
	local VERSION=$(git show upstream/${APPBND_BRANCH}:$1/app.properties | grep -F app.marketplace.version= | cut -d'=' -f 2)

	echo -n '      {'
	echo -n '"marketplace": "'${ID}'",'
	echo -n '"name": "'${NAME}'", '

	echo -n '"category": "'
	git show upstream/${APPBND_BRANCH}:${APPBND_FOLDER}/app.bnd | grep -F Liferay-Releng-Category | cut -d' ' -f 2- | tr -d '\n'
	echo -n '", '

	echo -n '"labs": '

	if [ "" == "$()" ]; then
		echo -n false
	else
		git show upstream/${APPBND_BRANCH}:${APPBND_FOLDER}/app.bnd | grep -F Liferay-Releng-Labs | cut -d' ' -f 2-
	fi

	echo -n ', '

	echo -n '"currentTags": ['

	grep -F "marketplace-$(basename ${MODULE_FOLDER})-${VERSION}-${LIFERAY_VERSION_NUMBER}" marketplace-tags.txt | \
		awk '{ print "\"" $1 "\"" }' | tr '\n' ',' | sed 's/,$//g'

	echo -n '], '

	echo -n '"pastTags": ['

	grep "marketplace-$(basename ${MODULE_FOLDER})-[0-9]*.[0-9]*.[0-9]*-${LIFERAY_VERSION_NUMBER}" marketplace-tags.txt | \
		grep -vF "marketplace-$(basename ${MODULE_FOLDER})-${VERSION}-${LIFERAY_VERSION_NUMBER}" | \
		awk '{ print "\"" $1 "\"" }' | tr '\n' ',' | sed 's/,$//g'

	echo -n '], '

	echo -n '"carolTags": ['

	grep -F "marketplace-$(basename ${MODULE_FOLDER})-${VERSION}-${LIFERAY_VERSION_NUMBER}" carol-tags.txt | \
		awk '{ print "\"" $1 "\"" }' | tr '\n' ',' | sed 's/,$//g'

	echo -n '], '

	echo -n '"version": "'${VERSION}'", '
	echo -n '"path": "'${MODULE_FOLDER}'", '

	local RELEASE_HASH=

	if [ "" != "$(grep -F ${1}/app.changelog .modules)" ]; then
		RELEASE_HASH=$(git show upstream/${APPBND_BRANCH}:$1/app.changelog | grep -F app.git.id-${VERSION}= | cut -d'=' -f 2)

		if [ "" == "${RELEASE_HASH}" ] || [ "" == "$(git log -1 --pretty='%H' ${RELEASE_HASH} 2> /dev/null)" ]; then
			RELEASE_HASH=$(git log -1 --pretty='%H' upstream/${APPBND_BRANCH} -- ${1}/app.changelog)
		fi

		if [ "${APPBND_FOLDER}" != "${MODULE_FOLDER}" ]; then
			RELEASE_HASH=$(git show ${RELEASE_HASH}:git-commit-portal-private 2> /dev/null)
		fi
	fi

	local CHANGELOG=

	if [ "" != "${RELEASE_HASH}" ] && [ "" != "$(git log -1 --pretty='%H' ${RELEASE_HASH} 2> /dev/null)" ]; then
		echo -n '"releaseDate": "'$(git log -1 --pretty='%ci' ${RELEASE_HASH} | awk '{ print $1 }')'", '
		CHANGELOG=$(git log --pretty='%s' ${RELEASE_HASH}..upstream/${MODULE_BRANCH} -- ${MODULE_FOLDER} | grep -Eo '[A-Z]+-[0-9]+' | grep -vF 'LRDOCS' | grep -vF LRQA | grep -vF -- '-0' | sort -u)
	else
		echo -n '"releaseDate": null, '
	fi

	echo -n '"fixes": ['

	if [ "" != "${CHANGELOG}" ]; then
		echo -n "${CHANGELOG}" | awk '{ print "\"" $1 "\"" }' | tr '\n' ',' | sed 's/,$//g' | sed 's/,/, /g'
	fi

	echo -n ']'
	echo '},'
}

if [ -f ${HOME}/source/release.properties ]; then
	cd ${HOME}/source
fi

if [ ! -f release.properties ]; then
	echo 'Please navigate to a portal source folder first'
	exit 1
fi

if git fetch upstream --no-tags; then
	git ls-remote -t git@github.com:liferay/liferay-portal-ee.git | awk '{ print $2 }' |
		grep '^refs/tags/marketplace-.*-[0-9]*10' | cut -d'/' -f 3 > marketplace-tags.txt


getparent.py
#!/usr/bin/env python

import git
from git import current_branch, git_root
import os
from os.path import abspath, dirname, isdir, isfile, join, relpath
import subprocess
import sys

def get_file_property(file_name, property):
	needle = '%s=' % property

	with open(file_name, 'r') as file:
		lines = file.readlines()
		lines = [line.strip() for line in lines if line.find(needle) > -1]
		return lines[0].split('=')[1] if len(lines) > 0 else None

def get_git_file_property(commit, file_name, property):
	lines = git.show('%s:%s' % (commit, file_name)).split('\n')
	lines = [line.strip() for line in lines if line.find(property) > -1]
	return lines[0].split('=')[1] if len(lines) > 0 else None

def getparent(check_tags):
	if git_root is None:
		return current_branch

	# Find the current branch, accounting for detached head

	ee_branches = ['6.2.x', '6.1.x', '6.0.x']
	ee_branches = ee_branches + ['ee-%s' % branch for branch in ee_branches]

	de_branches = ['7.0.x', '7.0.x-private', 'ee-7.0.x']

	dxp_branches = ['7.3.x', '7.2.x', '7.1.x']
	dxp_branches = dxp_branches + ['%s-private' % branch for branch in dxp_branches]

	if current_branch == 'master' or current_branch in ee_branches + de_branches + dxp_branches:
		return current_branch

	# Extract the full version

	full_version = None

	if isfile(join(git_root, 'release.properties')):
		full_version = get_file_property(join(git_root, 'release.properties'), 'lp.version')
	elif isfile(join(git_root, 'build.properties')) and isfile(join(git_root, 'app.server.properties')):
		full_version = get_file_property(join(git_root, 'build.properties'), 'lp.version')
	elif isfile(join(git_root, 'git-commit-portal')):
		with open(join(git_root, 'git-commit-portal'), 'r') as file:
			commit = file.readlines()[0].strip()
			full_version = get_git_file_property(commit, 'release.properties', 'lp.version')
	else:
		return getparent_origin()

	# If the short version is 6.x, then we have a shortcut

	short_version = '.'.join(full_version.split('.')[0:2])

	base_branch = None

	if short_version == '6.0':
		base_branch = 'ee-6.0.x'
	elif short_version == '6.1':
		base_branch = 'ee-6.1.x'
	elif short_version == '6.2':
		base_branch = 'ee-6.2.x'
	else:
		# Determine the base version using build.properties

		if isfile(join(git_root, 'build.properties')):
			base_branch = get_file_property(join(git_root, 'build.properties'), 'git.working.branch.name')

		if base_branch is None and isfile(join(git_root, 'git-commit-portal')):
			with open(join(git_root, 'git-commit-portal'), 'r') as file:
				commit = file.readlines()[0].strip()
				base_branch = get_git_file_property(commit, 'build.properties', 'git.working.branch.name')

		if base_branch is None:
			base_branch = current_branch if current_branch != 'HEAD' else '7.0.x'
		elif base_branch == 'ee-7.0.x':
			base_branch = '7.0.x'
		elif isdir(join(git_root, 'modules/private')) and len(git.ls_files('build.properties').strip()) == 0 and base_branch.find('-private') == -1:
			base_branch = '%s-private' % base_branch

	# If this is master or master-private, or we've recently rebased to 7.0.x or 7.0.x-private,
	# then use the branch instead of the tag

	if not check_tags:
		return base_branch

	# Find the closest matching tag

	base_tag = None

	if base_branch in dxp_branches or base_branch in de_branches:
		base_tag = git.describe('HEAD', '--tags', '--abbrev=0', '--match=fix-pack-*-%s%s10*' % (base_branch[0], base_branch[2]))

		if base_tag is None or len(base_tag) == 0:
			base_tag = git.describe('HEAD', '--tags', '--abbrev=0', '--match=%s.%s.*-ga*' % (base_branch[0], base_branch[2]))

	elif base_branch in ee_branches:
		base_tag = git.describe('HEAD', '--tags', '--abbrev=0', '--match=fix-pack-base-6%s*' % base_branch[5])

		if base_tag is None or len(base_tag) == 0:
			if base_branch.find('ee-') == 0:
				base_tag = git.describe('HEAD', '--tags', '--abbrev=0', '--match=%s.%s.*-ga*' % (base_branch[3], base_branch[5]))
			else:
				base_tag = git.describe('HEAD', '--tags', '--abbrev=0', '--match=%s.%s.*-ga*' % (base_branch[0], base_branch[2]))

	if base_tag is None:
		return base_branch

	if base_tag.find('fix-pack-base-') == 0 or base_tag.find('fix-pack-de-') == 0 or base_tag.find('fix-pack-dxp-') == 0 or base_tag.find('-ga') > -1:
		return base_tag

	return base_branch

def getparent_origin():
	remote_refs = git.for_each_ref('--format=%(refname)', 'refs/remotes/').split('\n')

	origin_branches = [ref[len('refs/remotes/'):] for ref in remote_refs if ref.find('refs/remotes/origin') == 0 and ref[-5:] != '/HEAD']
	upstream_branches = [ref[len('refs/remotes/'):] for ref in remote_refs if ref.find('refs/remotes/upstream') == 0 and ref[-5:] != '/HEAD']

	closest_branch = None
	closest_branch_diff = -1

	for branch_set in [upstream_branches, origin_branches]:
		for branch in branch_set:
			short_branch = branch[branch.find('/')+1:]

			if short_branch == current_branch or short_branch == closest_branch or not git.is_ancestor(branch, current_branch):
				continue

			branch_diff = len(git.log('--pretty=%H', '%s..%s' % (branch, current_branch)))

			if closest_branch is None or branch_diff < closest_branch_diff:
				closest_branch = short_branch
				closest_branch_diff = branch_diff

		if closest_branch is not None:
			return closest_branch

	return current_branch

if __name__ == '__main__':
	if len(sys.argv) == 1 or sys.argv[1] == 'tag':
		print(getparent(True))

	if len(sys.argv) == 1 or sys.argv[1] == 'branch':
		print(getparent(False))